crates/storage/provider/src/providers/database/builder.rs

```diff
@@ -0,0 +1,130 @@
+//! Helper builder entrypoint to instantiate a [`ProviderFactory`].
+//!
+//! This also includes general purpose staging types that provide builder style functions that lead
+//! up to the intended build target.
+
+use crate::{providers::StaticFileProvider, ProviderFactory};
+use reth_db_api::{
+    database_metrics::{DatabaseMetadata, DatabaseMetrics},
+    Database,
+};
+use reth_node_types::{NodeTypes, NodeTypesWithDBAdapter};
+use std::{marker::PhantomData, sync::Arc};
+
+/// Helper type to create a [`ProviderFactory`].
+///
+/// This type is the entry point for a stage based builder.
+///
+/// The intended staging is:
+///  1. Configure the database: [`ProviderFactoryBuilder::db`]
+///  2. Configure the chainspec: `chainspec`
+///  3. Configure the [`StaticFileProvider`]: `static_file`
+#[derive(Debug)]
+pub struct ProviderFactoryBuilder<N> {
+    _types: PhantomData<N>,
+}
+
+impl<N> ProviderFactoryBuilder<N> {
+    /// Maps the [`NodeTypes`] of this builder.
+    pub fn types<T>(self) -> ProviderFactoryBuilder<T> {
```

Comment:
How would this fn be used? Would it be used in combination with eg a constructor method on `ProviderFactoryBuilder<()>`

So the pattern would be something like
```rust
// Hopefully don't have to specify N for Default if planning on configuring types with the fn
let builder = ProviderFactoryBuilder::default()
    .types(types)
    .db(db)
    .chain_spec(spec)
    .static_files(sf)
    .build()
```

Reply:
yep,